### ISSUE TITLE
Move the translation between qemu-ir and llvm-bitcode from svm-node to vm-node

### ISSUE THREAD

**Issue by likebreath** [SVL-PSU/crete-dev#32]
In this way, the size of trace that needs to be transmitted (from vm-node to dispatch to svm-node) can be reduced dramatically, as the size of translation context normally took more than half of the whole captured trace.

**Comment by likebreath** [comment]
@moralismercatus My commit f97a1e034685e1f25665c1d4d08fa1cb65ea17fd does not tolerate any failures for the translation being moved into vm-node, such as when the translation failed, vm-node fsm can tolerate the failure and log the failed trace, etc. We need to talk about how to do that in a good manner. Please let me know your thoughts. Thanks.

**Comment by moralismercatus** [comment]
This looks good so far. It makes sense to do the translation in vm-node. I do wonder if the translation belongs in VMNodeFSM, though. Logically, yes, but practically I'm not sure. The reason is that subsequent QEMU execution is postponed while the translation is being done. If translation is very fast, it may not be a concern, but otherwise it may slow down the overall process.

Let me hear your opinion on translation speed before discussing fault tolerance.

Thanks,

**Comment by likebreath** [comment]
Yes, you are right. At the beginning, I thought the translation would not delay QEMU execution, as they are executed in different threads. But it seems they are synchronized by a "trace_ready" file, and in practical they are executed sequentially.

So far, my feeling is that the translation does not add obvious slowdown to the VMNodeFSM. Even though, we need to think about whether we want to put the translation as a part of VMNodeFSM. If we use the same fault tolerance on the translation in VMNode like what we did in SVMNode, we might need to reset the QEMU when translation has an error, which I don't think is a good idea. This is also why I did not do it in my last commit. Please let me know your thoughts. Thanks.

**Comment by moralismercatus** [comment]
Yes, if translation fails in VMNodeFSM, the whole FSM will reset including the QEMU instance. So, that's two shortcomings of placing it there: 1) delays next QEMU test; 2) unnecessarily resets QEMU/FSM on translation failure.

For, (1) keeping translation as part of the FSM, I think the delay is unavoidable, if we are to keep to the original design goals of the FSM. Each task must be completed before transitioning to the next state. As for (2), there are ways to avoid resetting the whole FSM, since this is done by VMNode. One way would be to throw a special exception from the state action on translation failure, and, handle it specially in the exception_caught() function. It's a little tricky though. I had experimented with a VMRecoverableException that transitions to a RecoverableError state, but that's adding complexity.

It really seems simpler to me to maintain a vector/pool of AsyncTask in VMNode. In poll(), for the flag::trace_ready clause, launch translate_trace as an AsyncTask. Then, at each poll, also check async_tasks if any traces are ready yet, and only then push them to the trace queue. Something like that, if that makes sense.

I could write up an impl. for either of these strategies fairly quickly, I think, if you'd like.

**Comment by likebreath** [comment]
Closing this one. Handling of translation fails in VMNodeFSM will be discussed in a separate issue.
